### PR TITLE
Fix type hints

### PR DESCRIPTION
**Pull request by kerizane**
Did some work on #1211, this is it for now.

### CODE REVIEW

File: tuxemon/db.py
Comment (vnmabus) [inline tuxemon/db.py]: Why did you remove the return value?

**Review by vnmabus — COMMENTED** [review]
I also have another suggestion. In Python 3.10 the `|` operator has been overloaded to create union of types (https://peps.python.org/pep-0604/). Mypy supports the syntax also for previous Python versions if the evaluation of the annotations is postponed (with `from __future__ import annotations`). As the syntax is shorter and way clearer, I propose to use it in new type annotations (or when old modules are touched). Thus, instead of `Optional[int]` one would write `int | None`, or instead of `Union[int, str]` one would write `int | str`. Note that this is only for annotations: definition of types or aliases still need to use `Optional` and `Union` until we remove support for Python versions prior to 3.10.

File: tuxemon/db.py
Comment (kerizane) [reply tuxemon/db.py]: The examples in the PEP that introduced overload didn't have typehints on the actual function, I assumed that was how it worked.

However, I can see that mypy wants it to be typehinted; I will fix it.

https://mypy.readthedocs.io/en/stable/more_types.html#function-overloading

Comment (kerizane) [reply tuxemon/db.py]: Just read the article more carefully... typehints are optional on the implementation. We may as well keep them.

Comment (kerizane) [reply tuxemon/db.py]:
```
    @overload
    def lookup(
        self,
        slug: str,
        table: Literal["environment"],
    ) -> EnvironmentModel:
        pass

    @overload
    def lookup(
        self,
        slug: str,
        table: str,
    ) -> Mapping[str, Any]:
        pass
```
Sigh... mypy doesn't understand this, even though this behaviour is explained in the PEP https://peps.python.org/pep-0586/#interactions-with-overloads

"Overloaded function signatures 11 and 12 overlap with incompatible return types"

Comment (kerizane) [reply tuxemon/db.py:731]: Simple reproduction of the mypy error:

```
from typing import Literal, Union, overload


@overload
def foo(s: Literal["hi"]) -> int:
    ...


@overload
def foo(s: str) -> str:
    ...


def foo(s: str) -> Union[str, int]:
    if s == "hi":
        return 1
    return s
```
example.py:5: error: Overloaded function signatures 1 and 2 overlap with incompatible return types

___

Will skip this one for now.

Comment (kerizane) [reply tuxemon/db.py]: Actually, I see what's going on here. The fallback option needs to be a superset of the other returning types, there's no option for "string that isn't one of the previous literals"

Comment (vnmabus) [reply tuxemon/db.py]: The REAL problem is that since  #1192 landed, the returning type is not a `Mapping[str, Any]`, but a pydantic `BaseModel`. That is the "incompatible return types" Mypy complains about.

Comment (kerizane) [reply tuxemon/db.py]: There's still one branch where it is a Mapping[str, Any], so it should be a Union of BaseModel and Mapping[str, Any].

https://github.com/Tuxemon/Tuxemon/blob/1c9b59686679a4ca7685027292e0881d1545a55d/tuxemon/db.py#L682

The solution is to either remove this branch, if it's unused, or to make the returning type a union of all the above Models (or just BaseModel) and Mapping[str, Any].

Comment (vnmabus) [reply tuxemon/db.py]: That should be fixed then, to have a common base class.

Comment (kerizane) [reply tuxemon/db.py]: Is the common base class just for the existing models, or do you want the Mapping[str, Any] to be included somehow?

I guess for unknown tables we could wrap the dict in a Model. I don't see much point, except it makes the typehints slightly more concise. We also shouldn't have unknown tables, so don't see much point in keeping this branch around. I guess it could be useful for development when you're implementing a new table and want to use a dict at first, rather than defining the model straight away.

```
class DataModel(BaseModel):
    data: Mapping[str, Any]
```
(Didn't check the syntax, but you know what I mean)

**Comment by kerizane** [comment]
> I also have another suggestion. In Python 3.10 the `|` operator has been overloaded to create union of types (https://peps.python.org/pep-0604/). Mypy supports the syntax also for previous Python versions if the evaluation of the annotations is postponed (with `from __future__ import annotations`). As the syntax is shorter and way clearer, I propose to use it in new type annotations (or when old modules are touched). Thus, instead of `Optional[int]` one would write `int | None`, or instead of `Union[int, str]` one would write `int | str`. Note that this is only for annotations: definition of types or aliases still need to use `Optional` and `Union` until we remove support for Python versions prior to 3.10.

Yeah, that would be nice. The only issue is that the inconsistency might be confusing for some people: e.g. newcomers or people who haven't used typehints much.

I'll save this for another ticket, unless somebody like @bitcraft agrees with you.

**Comment by kerizane** [comment]
Tested the game, seems fine.

**Comment by kerizane** [comment]
Turns out we actually started with 365 (I didn't have all the libraries installed when I did my first scan), I've brought it down to 172.

This will probably do for now.

File: tuxemon/cli/processor.py
Comment (vnmabus) [inline tuxemon/cli/processor.py]: Why create it empty and then append a single folder?

File: tuxemon/cli/commands/whereami.py
Comment (vnmabus) [inline tuxemon/cli/commands/whereami.py]: Why the `try` now?

File: mypy.ini
Comment (vnmabus) [inline mypy.ini:3]:
```suggestion
ignore_missing_imports = true
enable_error_code = ignore-without-code
```

File: tuxemon/db.py
Comment (vnmabus) [inline tuxemon/db.py]: Is that mapping REALLY needed? Also, can we use a base class instead of the full list?

File: tuxemon/event/actions/start_cinema_mode.py
Comment (vnmabus) [inline tuxemon/event/actions/start_cinema_mode.py]: `assert` is better here. Works for type checking and fails at runtime.

File: tuxemon/event/actions/stop_cinema_mode.py
Comment (vnmabus) [inline tuxemon/event/actions/stop_cinema_mode.py]: `assert` again.

File: tuxemon/save.py
Comment (vnmabus) [inline tuxemon/save.py]: Maybe we should use a different type to prevent mistakes?

File: tuxemon/states/world/worldstate.py
Comment (vnmabus) [inline tuxemon/states/world/worldstate.py]: Do we really want to allow modifications?

File: tuxemon/tools.py
Comment (vnmabus) [inline tuxemon/tools.py]: This change is wrong, there is no warning anymore.

**Review by vnmabus — CHANGES_REQUESTED** [review]

**Comment by kerizane** [comment]
Thanks for the suggestions, they all look good.

File: tuxemon/tools.py
Comment (kerizane) [reply tuxemon/tools.py]: https://peps.python.org/pep-0484/#the-noreturn-type

NoReturn shouldn't be used for arguments. I will doublecheck what this function is doing, but it should probably be Any.

Comment (vnmabus) [reply tuxemon/tools.py]: That function is used to detect that every possible value of that variable has been handled at that point. That is why the type is `NoReturn` (which really means "no type allowed").

Comment (kerizane) [reply tuxemon/tools.py]: I get this error when using mypy 0.761 on the original code:

> tuxemon/states/combat/combat.py:353: error: Argument 1 to "assert_never" has incompatible type "Union[Literal['begin'], Literal['ready'], Literal['housekeeping phase'], Literal['decision phase'], Literal['pre action phase'], Literal['action phase'], Literal['post action phase'], Literal['resolve match'], Literal['ran away'], Literal['draw match'], Literal['has winner'], Literal['end combat']]"; expected "NoReturn"

NoReturn as a return type makes perfect sense, but you shouldn't use NoReturn as an annotation for "value".

See https://peps.python.org/pep-0484/#the-noreturn-type

> The NoReturn type is only valid as a return annotation of functions, and considered an error if it appears in other positions:

```
from typing import List, NoReturn

# All of the following are errors
def bad1(x: NoReturn) -> int:
    ...
bad2 = None  # type: NoReturn
def bad3() -> List[NoReturn]:
```
___
The signature should be `def assert_never(value: Any) -> NoReturn:` as you want to be able to use assert_never on variables of any type.

Comment (vnmabus) [reply tuxemon/tools.py]: Then it means that these literals have not been handled. If you change the type, the function is useless.

There is a similar function in Python 3.11, but we cannot use it yet.

Comment (kerizane) [reply tuxemon/tools.py:385]:
```
from typing import Any, Literal, NoReturn

CombatPhase = Literal[
    "begin",
]


def assert_never(value: NoReturn) -> NoReturn:
    assert False


def outer(phase: CombatPhase):
    if phase == "begin":
        return
    else:
        assert_never(phase)
```

example.py:16: error: Argument 1 to "assert_never" has incompatible type "Literal['begin']"; expected "NoReturn"

Comment (kerizane) [reply tuxemon/tools.py]: Oh, you're right.

https://typing.readthedocs.io/en/latest/source/unreachable.html

It seems this was originally a bit of a hack (https://github.com/python/mypy/issues/5818), which is why this surprised me.

It turns out I had a very old version of mypy installed which didn't understand the idiom. Sorry about that.

File: tuxemon/save.py
Comment (kerizane) [reply tuxemon/save.py]: What do you mean by this?

Comment (vnmabus) [reply tuxemon/save.py]: Using `NewType` for it, to prevent mistakenly assigning a normal string.

File: tuxemon/db.py
Comment (kerizane) [reply tuxemon/db.py]: I don't think it makes sense to add a base class, as we don't use these models polymorphically.

I could make an alias for the union if you think it's too verbose.

File: tuxemon/map_loader.py
Comment (vnmabus) [inline tuxemon/map_loader.py]: I would prefer not to cast. Instead, we should properly type `pytmx`.

File: tuxemon/monster.py
Comment (vnmabus) [inline tuxemon/monster.py]: Why is the cast needed? In case that is necessary, I would prefer an `assert`.

File: tuxemon/save_upgrader.py
Comment (vnmabus) [inline tuxemon/save_upgrader.py:68]: Why isn't the input a `SaveData` too?

File: tuxemon/states/combat/combat_animations.py
Comment (vnmabus) [inline tuxemon/states/combat/combat_animations.py:75]: That should be fixed by doing what is proposed in #1167.

**Review by vnmabus — CHANGES_REQUESTED** [review]

File: tuxemon/map_loader.py
Comment (kerizane) [reply tuxemon/map_loader.py]: Agreed

Comment (kerizane) [reply tuxemon/map_loader.py]: What about natsorted? Their typehints are also bad.

File: tuxemon/monster.py
Comment (kerizane) [reply tuxemon/monster.py]:
```
class EnqueuedAction(NamedTuple):
    user: Union[Monster, NPC, None]
    technique: Union[Technique, Item]
    target: Monster
```

File: tuxemon/save_upgrader.py
Comment (kerizane) [reply tuxemon/save_upgrader.py:68]: save_upgrader converts a save file from an earlier format (i.e before refactoring) to the current format. The data that comes in might not match the schema of `SaveData`. Look at `_transfer_storage_boxes` within the same file for an example.

Comment (vnmabus) [reply tuxemon/save_upgrader.py:68]: Maybe then we should check that all fields are present and with appropriate types using a `TypeGuard`? We could use it if we add `typing_extensions` as a dependency, at least for type-checking.

File: tuxemon/monster.py
Comment (vnmabus) [reply tuxemon/monster.py]: An `assert` then, better than a cast in my opinion.

File: tuxemon/map_loader.py
Comment (vnmabus) [reply tuxemon/map_loader.py]: Well, I said that about `pytmx` because its author is a maintainer of this project, so they will certainly be accepted. If we can improve `natsort` we should  also try to, but I think the last commit is from 7 months ago, so maybe the PR will be stalled for a while.

File: tuxemon/states/combat/combat_animations.py
Comment (kerizane) [reply tuxemon/states/combat/combat_animations.py:75]: Yes, looking forward to https://github.com/Tuxemon/Tuxemon/issues/1167 being done.

Do you have a suggestion related to this part of the code, or was it just an FYI? I think it's reasonable to explain that the existing typehints are wrong.

File: tuxemon/map_loader.py
Comment (kerizane) [reply tuxemon/map_loader.py]: Are you happy with me leaving the natsorted cast in? Bitcraft is busy atm (I think), so don't know how quickly a pytmx fix will be accepted. But I'm still happy to remove the pytmx cast.

File: tuxemon/save_upgrader.py
Comment (kerizane) [reply tuxemon/save_upgrader.py:68]: ~I'd rather not make this PR too complicated. Could you raise an issue explaining your idea?~

Actually, I don't think that's too bad. Maybe I will do it.

File: tuxemon/monster.py
Comment (kerizane) [reply tuxemon/monster.py]: Done.

File: tuxemon/map_loader.py
Comment (kerizane) [reply tuxemon/map_loader.py]: I'm not sure what to do here. natsorted gives a stupid error, but it feels slightly silly adding a pointless assert just to stop mypy complaining.

tuxemon/map_loader.py:395: error: Item "float" of "Union[str, bytes, float, int, PurePath, None, Iterable[Union[str, bytes, float, int, PurePath, None, Iterable[Any]]]]" has no attribute "startswith"
tuxemon/map_loader.py:395: error: Item "int" of "Union[str, bytes, float, int, PurePath, None, Iterable[Union[str, bytes, float, int, PurePath, None, Iterable[Any]]]]" has no attribute "startswith"
tuxemon/map_loader.py:395: error: Item "PurePath" of "Union[str, bytes, float, int, PurePath, None, Iterable[Union[str, bytes, float, int, PurePath, None, Iterable[Any]]]]" has no attribute "startswith"
tuxemon/map_loader.py:395: error: Item "None" of "Union[str, bytes, float, int, PurePath, None, Iterable[Union[str, bytes, float, int, PurePath, None, Iterable[Any]]]]" has no attribute "startswith"
tuxemon/map_loader.py:395: error: Item "Iterable[Union[str, bytes, float, int, PurePath, None, Iterable[Any]]]" of "Union[str, bytes, float, int, PurePath, None, Iterable[Union[str, bytes, float, int, PurePath, None, Iterable[Any]]]]" has no attribute "startswith"

File: tuxemon/save_upgrader.py
Comment (kerizane) [reply tuxemon/save_upgrader.py:68]: We could also use pydantic.

File: tuxemon/states/combat/combat_animations.py
Comment (vnmabus) [reply tuxemon/states/combat/combat_animations.py:75]: Just a FYI.

Comment (kerizane) [reply tuxemon/states/combat/combat_animations.py:75]: I could just fix the typehint to make it accurate, even if that makes mypy complain.

It might not be possible though, I've seen exceptions in similar situations.

Comment (vnmabus) [reply tuxemon/states/combat/combat_animations.py:75]: It is fine to let it as is. We will tackle the root of the issue in #1167.

File: tuxemon/save_upgrader.py
Comment (kerizane) [reply tuxemon/save_upgrader.py:68]: Thought about this some more:

I've chosen to keep this PR simple, and tried to avoid functional changes. I've added some asserts, but they were generally asserting that values that can't be none weren't none, in places where the game would immediately crash anyway.

Validating the save file is a bigger change, and there is some risk to doing it. I'd rather not break somebody's save file just because of some irrelevant type error.

It is a good idea, but I'd prefer to isolate any risk and any technical discussions from this PR.

If you're still interested then please raise a feature request.

**Comment by bitcraft** [comment]
wow, lots of changes.  and its getting old at this point.  i see there are conflicts a different merge here.  in the interest of getting it merged quickly, we can leave the `|` operator changes for later if you prefer.  3.10 is the default download so we should feel free to assume people can/will use that.  as far as the pytmx changes, lmk what i can do to get that compliant.

**Comment by kerizane** [comment]
> as far as the pytmx changes, lmk what i can do to get that compliant.

The method is typed as returning `dict`, which gets interpreted as Dict[Any, Any]. Can you make it more specific? The key is always a str (I think) and the value is a union of types.

**Comment by kerizane** [comment]
I will fix the conflict and see where I got to with this PR. I will do the Union | None in another ticket, should be easy as pyupgrade can do it.

@vnmabus anything that needs before we can merge? Let's try to close this down.

**Comment by vnmabus** [comment]
I don't think so. It is fine for now.

**Comment by kerizane** [comment]
@vnmabus Thanks for all your help with this.

**Comment by kerizane** [comment]
@bitcraft Ready to go

**Review by bitcraft — APPROVED** [review]

**Pull request merged**
